components/script/dom/xrsession.rs

```diff
@@ -1070,7 +1070,7 @@ impl XRSessionMethods for XRSession {
             }),
         );
 
-        session.update_frame_rate(*rate, sender);
+        self.session.borrow_mut().update_frame_rate(*rate, sender);
```

Comment:
This won't work—the `self.session.borrow()` is still active, so this will panic. Unfortunately, there do not appear to be any tests that call this method :(

Reply:
Oh noo, how will I check if the implementation is working as expected then?

Reply:
I found a solution! I verified that running https://cityslingvr.arvr.social/ in Servo and clicking "click here to enter VR" causes UpdateTargetFrameRate to be called. Can you check if the browser crashes with your current changes applied?

Reply:
Great, I'll get back to you on this

Reply:
> I found a solution! I verified that running https://cityslingvr.arvr.social/ in Servo and clicking "click here to enter VR" causes UpdateTargetFrameRate to be called. Can you check if the browser crashes with your current changes applied?

Okay, I've tried this and there don't seem to be any crashes

Reply:
What happens when you press the button?

Reply:
Nothing, actually :/

Reply:
Anything interesting appear in the terminal output from servo?

Reply:
From my own testing, you may need to click the button several times to see an effect. Also, are you running on macOS?

Reply:
So, I've come across two cases:
1. After clicking on "Click here to enter VR", I can enter the VR mode but then a few moments later, the screen goes black (which I'm guessing is the "crash" you were referring to)
2. I get this error every time I click on it:

![Screenshot 2024-10-21 233307](https://github.com/user-attachments/assets/9482aeb8-2bb1-43d3-a48c-226de6c5b595)
And I'm running in a linux env

Reply:
Ok, I have a testing suggestion which "worked" when I applied your code changes:
* to ensure the implementation of UpdateTargetFrameRate gets to run, change `let supported_frame_rates = session.supported_frame_rates()` to `let supported_frame_rates = vec![*rate];`
* run `./mach run https://cityslingvr.arvr.social/`
* click "click here to enter VR"

I get the following panic, which is caused by the current code changes:
```
already borrowed: BorrowMutError (thread Script(1,1), at components/script/dom/xrsession.rs:1074)
   0: backtrace::backtrace::libunwind::trace
             at /Users/jdm/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.74/src/backtrace/libunwind.rs:116:5
      backtrace::backtrace::trace_unsynchronized
             at /Users/jdm/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.74/src/backtrace/mod.rs:66:5
   1: <servoshell::backtrace::Print as core::fmt::Debug>::fmt
             at /Users/jdm/src/servo/ports/servoshell/backtrace.rs:65:13
   2: core::fmt::rt::Argument::fmt
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/fmt/rt.rs:165:63
      core::fmt::write
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/fmt/mod.rs:1168:21
   3: std::io::Write::write_fmt
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/io/mod.rs:1835:15
   4: servoshell::backtrace::print
             at /Users/jdm/src/servo/ports/servoshell/backtrace.rs:18:5
   5: servoshell::panic_hook::panic_hook
             at /Users/jdm/src/servo/ports/servoshell/panic_hook.rs:40:17
   6: core::ops::function::Fn::call
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/ops/function.rs:79:5
   7: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/alloc/src/boxed.rs:2077:9
      std::panicking::rust_panic_with_hook
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panicking.rs:799:13
   8: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panicking.rs:664:13
   9: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/sys_common/backtrace.rs:171:18
  10: rust_begin_unwind
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panicking.rs:652:5
  11: core::panicking::panic_fmt
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/panicking.rs:72:14
  12: core::cell::panic_already_borrowed
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/cell.rs:785:5
  13: core::cell::RefCell<T>::borrow_mut
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/cell.rs:1076:25
  14: script::dom::bindings::cell::DomRefCell<T>::borrow_mut
             at /Users/jdm/src/servo/components/script/dom/bindings/cell.rs:121:9
  15: <script::dom::xrsession::XRSession as script::dom::bindings::codegen::Bindings::XRSessionBinding::XRSession_Binding::XRSessionMethods>::UpdateTargetFrameRate
             at /Users/jdm/src/servo/components/script/dom/xrsession.rs:1074:9
  16: script::dom::bindings::codegen::Bindings::XRSessionBinding::XRSession_Binding::updateTargetFrameRate::{{closure}}::{{closure}}
             at /Users/jdm/src/servo/target/debug/build/script-17526dfa1f56939a/out/Bindings/XRSessionBinding.rs:667:35
  17: script::dom::bindings::codegen::Bindings::XRSessionBinding::XRSession_Binding::updateTargetFrameRate::{{closure}}
             at /Users/jdm/src/servo/target/debug/build/script-17526dfa1f56939a/out/Bindings/XRSessionBinding.rs:647:33
  18: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/ops/function.rs:305:13
  19: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/panic/unwind_safe.rs:272:9
  20: std::panicking::try::do_call
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panicking.rs:559:40
  21: ___rust_try
  22: std::panicking::try
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panicking.rs:523:19
  23: std::panic::catch_unwind
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panic.rs:149:14
  24: mozjs::panic::wrap_panic
             at /Users/jdm/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/a02aaf1/mozjs/src/panic.rs:22:11
  25: script::dom::bindings::codegen::Bindings::XRSessionBinding::XRSession_Binding::updateTargetFrameRate
             at /Users/jdm/src/servo/target/debug/build/script-17526dfa1f56939a/out/Bindings/XRSessionBinding.rs:647:5
  26: script::dom::bindings::codegen::Bindings::XRSessionBinding::XRSession_Binding::updateTargetFrameRate_promise_wrapper::{{closure}}::{{closure}}
             at /Users/jdm/src/servo/target/debug/build/script-17526dfa1f56939a/out/Bindings/XRSessionBinding.rs:678:18
  27: script::dom::bindings::codegen::Bindings::XRSessionBinding::XRSession_Binding::updateTargetFrameRate_promise_wrapper::{{closure}}
             at /Users/jdm/src/servo/target/debug/build/script-17526dfa1f56939a/out/Bindings/XRSessionBinding.rs:677:33
  28: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/ops/function.rs:305:13
  29: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/panic/unwind_safe.rs:272:9
  30: std::panicking::try::do_call
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panicking.rs:559:40
  31: ___rust_try
  32: std::panicking::try
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panicking.rs:523:19
  33: std::panic::catch_unwind
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panic.rs:149:14
  34: mozjs::panic::wrap_panic
             at /Users/jdm/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/a02aaf1/mozjs/src/panic.rs:22:11
  35: script::dom::bindings::codegen::Bindings::XRSessionBinding::XRSession_Binding::updateTargetFrameRate_promise_wrapper
             at /Users/jdm/src/servo/target/debug/build/script-17526dfa1f56939a/out/Bindings/XRSessionBinding.rs:677:5
  36: CallJitMethodOp
             at /Users/runner/work/mozjs/mozjs/mozjs-sys/src/jsglue.cpp:638:10
  37: script::dom::bindings::utils::generic_call
             at /Users/jdm/src/servo/components/script/dom/bindings/utils.rs:522:5
  38: script::dom::bindings::utils::generic_method
             at /Users/jdm/src/servo/components/script/dom/bindings/utils.rs:538:5
  39: __ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
  40: __ZN2js9InterpretEP9JSContextRNS_8RunStateE
  41: __ZN2js9RunScriptEP9JSContextRNS_8RunStateE
  42: __ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
  43: __ZN2js4CallEP9JSContextN2JS6HandleINS2_5ValueEEES5_RKNS_13AnyInvokeArgsENS2_13MutableHandleIS4_EENS_10CallReasonE
  44: __Z20JS_CallFunctionValueP9JSContextN2JS6HandleIP8JSObjectEENS2_INS1_5ValueEEERKNS1_16HandleValueArrayENS1_13MutableHandleIS6_EE
  45: mozjs::rust::wrappers::JS_CallFunctionValue
             at /Users/jdm/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/a02aaf1/mozjs/src/rust.rs:1115:51
  46: script::dom::bindings::codegen::Bindings::EventListenerBinding::EventListener::HandleEvent
             at /Users/jdm/src/servo/target/debug/build/script-17526dfa1f56939a/out/Bindings/EventListenerBinding.rs:68:18
  47: script::dom::bindings::codegen::Bindings::EventListenerBinding::EventListener::HandleEvent_
             at /Users/jdm/src/servo/target/debug/build/script-17526dfa1f56939a/out/Bindings/EventListenerBinding.rs:36:18
  48: script::dom::eventtarget::CompiledEventListener::call_or_handle_event
             at /Users/jdm/src/servo/components/script/dom/eventtarget.rs:189:25
  49: script::dom::event::inner_invoke
             at /Users/jdm/src/servo/components/script/dom/event.rs:734:9
  50: script::dom::event::invoke
             at /Users/jdm/src/servo/components/script/dom/event.rs:666:17
  51: script::dom::event::Event::dispatch
             at /Users/jdm/src/servo/components/script/dom/event.rs:289:17
  52: script::dom::eventtarget::EventTarget::dispatch_event
             at /Users/jdm/src/servo/components/script/dom/eventtarget.rs:406:9
  53: <script::dom::eventtarget::EventTarget as script::dom::bindings::codegen::Bindings::EventTargetBinding::EventTarget_Binding::EventTargetMethods>::DispatchEvent
             at /Users/jdm/src/servo/components/script/dom/eventtarget.rs:799:18
  54: script::dom::bindings::codegen::Bindings::EventTargetBinding::EventTarget_Binding::dispatchEvent::{{closure}}::{{closure}}
             at /Users/jdm/src/servo/target/debug/build/script-17526dfa1f56939a/out/Bindings/EventTargetBinding.rs:367:43
  55: script::dom::bindings::codegen::Bindings::EventTargetBinding::EventTarget_Binding::dispatchEvent::{{closure}}
             at /Users/jdm/src/servo/target/debug/build/script-17526dfa1f56939a/out/Bindings/EventTargetBinding.rs:342:33
  56: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/ops/function.rs:305:13
  57: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/panic/unwind_safe.rs:272:9
  58: std::panicking::try::do_call
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panicking.rs:559:40
  59: ___rust_try
  60: std::panicking::try
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panicking.rs:523:19
  61: std::panic::catch_unwind
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panic.rs:149:14
  62: mozjs::panic::wrap_panic
             at /Users/jdm/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/a02aaf1/mozjs/src/panic.rs:22:11
  63: script::dom::bindings::codegen::Bindings::EventTargetBinding::EventTarget_Binding::dispatchEvent
             at /Users/jdm/src/servo/target/debug/build/script-17526dfa1f56939a/out/Bindings/EventTargetBinding.rs:342:5
  64: CallJitMethodOp
             at /Users/runner/work/mozjs/mozjs/mozjs-sys/src/jsglue.cpp:638:10
  65: script::dom::bindings::utils::generic_call
             at /Users/jdm/src/servo/components/script/dom/bindings/utils.rs:522:5
  66: script::dom::bindings::utils::generic_method
             at /Users/jdm/src/servo/components/script/dom/bindings/utils.rs:538:5
```

Reply:
To be clear, changing the set of supported frame rates is only to allow you to test your code changes; we don't want to actually make that change permanently.

Reply:
Understood, let me look into this 👍